File: webrtc_surgical_platform/external_config.py
#!/usr/bin/env python3
"""
External Configuration Manager for WebRTC Bridge
Handles dynamic configuration for local vs external (Ngrok) deployments
"""

import os
import json
from typing import Dict, Optional
from dotenv import load_dotenv


class ExternalConfigManager:
    """Manages external configuration for WebRTC bridge service"""
    
    def __init__(self):
        self.config = {}
        self.load_configuration()
    
    def load_configuration(self):
        """Load configuration from environment files"""
        # Load base .env file first
        base_env_path = '.env'
        if os.path.exists(base_env_path):
            load_dotenv(base_env_path)
        
        # Check if external mode is enabled
        external_mode = (
            os.getenv('EXTERNAL_MODE', 'false').lower() == 'true' or 
            os.getenv('NODE_ENV') == 'external'
        )
        
        if external_mode:
            # Load external configuration
            external_env_path = '.env.external'
            if os.path.exists(external_env_path):
                load_dotenv(external_env_path, override=True)
                print('🌐 Python external configuration loaded')
            else:
                print('⚠️  External mode enabled but .env.external not found')
        
        # Build unified configuration
        self.config = {
            # External Mode Settings
            'external_mode': external_mode,
            'is_external': external_mode,
            
            # Server Configuration
            'webrtc_backend_url': os.getenv('API_BASE_URL', 'http://localhost:3001'),
            'backend_external_url': os.getenv('BACKEND_EXTERNAL_URL', None),
            
            # Bridge Configuration  
            'bridge_websocket_url': os.getenv('AR_BRIDGE_URL', 'ws://localhost:8765'),
            'bridge_http_url': os.getenv('AR_BRIDGE_HTTP_URL', 'http://localhost:8766'),
            'bridge_ws_port': int(os.getenv('AR_BRIDGE_WS_PORT', '8765')),
            'bridge_http_port': int(os.getenv('AR_BRIDGE_HTTP_PORT', '8766')),
            
            # Frontend Configuration
            'frontend_url': os.getenv('FRONTEND_EXTERNAL_URL', 'http://localhost:3000'),
            'cors_origins': self._build_cors_origins(),
            
            # External-specific settings
            'external_timeout': int(os.getenv('EXTERNAL_TIMEOUT', '60000')),
            'connection_retry_attempts': int(os.getenv('CONNECTION_RETRY_ATTEMPTS', '3')),
            'heartbeat_interval': int(os.getenv('HEARTBEAT_INTERVAL', '30000')),
            
            # Ngrok settings
            'ngrok_auth_token': os.getenv('NGROK_AUTH_TOKEN'),
            'ngrok_region': os.getenv('NGROK_REGION', 'us'),
        }
    
    def _build_cors_origins(self) -> list:
        """Build CORS origins list"""
        origins = [
            os.getenv('CORS_ORIGIN', 'http://localhost:3000')
        ]
        
        # Add external URL if available
        if os.getenv('FRONTEND_EXTERNAL_URL'):
            origins.append(os.getenv('FRONTEND_EXTERNAL_URL'))
        
        # Add Ngrok patterns for external mode
        if self.config.get('external_mode', False):
            origins.extend([
                'https://*.ngrok-free.app',
                'https://*.ngrok.io',
                'https://*.ngrok.app'
            ])
        
        return list(set(origins))  # Remove duplicates
    
    def update_external_urls(self, urls: Dict[str, str]):
        """
        Update configuration with dynamically generated Ngrok URLs
        
        Args:
            urls: Dictionary containing external URLs
                  {
                      'backend': 'https://abc.ngrok-free.app',
                      'frontend': 'https://def.ngrok-free.app', 
                      'bridge': 'wss://ghi.ngrok-free.app',
                      'bridge_http': 'https://jkl.ngrok-free.app'
                  }
        """
        if 'backend' in urls:
            self.config['webrtc_backend_url'] = urls['backend']
            self.config['backend_external_url'] = urls['backend']
            os.environ['API_BASE_URL'] = urls['backend']
            os.environ['BACKEND_EXTERNAL_URL'] = urls['backend']
        
        if 'frontend' in urls:
            self.config['frontend_url'] = urls['frontend']
            os.environ['FRONTEND_EXTERNAL_URL'] = urls['frontend']
            os.environ['CORS_ORIGIN'] = urls['frontend']
        
        if 'bridge' in urls:
            self.config['bridge_websocket_url'] = urls['bridge']
            os.environ['AR_BRIDGE_URL'] = urls['bridge']
        
        if 'bridge_http' in urls:
            self.config['bridge_http_url'] = urls['bridge_http']
            os.environ['AR_BRIDGE_HTTP_URL'] = urls['bridge_http']
        
        # Update CORS origins
        self.config['cors_origins'] = self._build_cors_origins()
        
        print('🔄 Python external URLs updated:', {
            'backend': self.config['webrtc_backend_url'],
            'frontend': self.config['frontend_url'],
            'bridge_ws': self.config['bridge_websocket_url'],
            'bridge_http': self.config['bridge_http_url']
        })
    
    def get_config(self) -> Dict:
        """Get current configuration"""
        return self.config.copy()
    
    def is_external_mode(self) -> bool:
        """Check if running in external mode"""
        return self.config.get('external_mode', False)
    
    def get_webrtc_backend_url(self) -> str:
        """Get WebRTC backend URL"""
        return self.config['webrtc_backend_url']
    
    def get_bridge_websocket_url(self) -> str:
        """Get bridge WebSocket URL"""
        return self.config['bridge_websocket_url']
    
    def get_bridge_http_url(self) -> str:
        """Get bridge HTTP URL"""
        return self.config['bridge_http_url']
    
    def get_frontend_url(self) -> str:
        """Get frontend URL"""
        return self.config['frontend_url']
    
    def get_cors_origins(self) -> list:
        """Get CORS origins"""
        return self.config['cors_origins']
    
    def get_display_info(self) -> Dict:
        """Get configuration display info"""
        mode = 'External' if self.is_external_mode() else 'Local'
        return {
            'mode': mode,
            'backend': self.config['webrtc_backend_url'],
            'frontend': self.config['frontend_url'],
            'bridge_ws': self.config['bridge_websocket_url'],
            'bridge_http': self.config['bridge_http_url'],
            'cors_origins': self.config['cors_origins']
        }
    
    def log_configuration(self):
        """Log current configuration"""
        config_info = self.get_display_info()
        print('🔧 Python Bridge Configuration:', json.dumps(config_info, indent=2))
        
        if self.is_external_mode():
            print('🌐 Python bridge running in EXTERNAL mode - ready for Ngrok tunnels')
        else:
            print('🏠 Python bridge running in LOCAL mode')


# Create singleton instance
external_config = ExternalConfigManager()

# Log configuration on import
external_config.log_configuration()

File: webrtc_surgical_platform/backend/src/config/externalConfig.js
/**
 * External Configuration Manager
 * Handles dynamic configuration for local vs external (Ngrok) deployments
 */

const path = require('path');
const fs = require('fs');
const dotenv = require('dotenv');

class ExternalConfigManager {
    constructor() {
        this.config = {};
        this.loadConfiguration();
    }

    loadConfiguration() {
        // Load base .env file first
        const baseEnvPath = path.join(process.cwd(), '.env');
        if (fs.existsSync(baseEnvPath)) {
            dotenv.config({ path: baseEnvPath });
        }

        // Check if external mode is enabled
        const externalMode = process.env.EXTERNAL_MODE === 'true' || 
                           process.env.NODE_ENV === 'external';

        if (externalMode) {
            // Load external configuration
            const externalEnvPath = path.join(process.cwd(), '.env.external');
            if (fs.existsSync(externalEnvPath)) {
                dotenv.config({ path: externalEnvPath, override: true });
                console.log('🌐 External configuration loaded');
            } else {
                console.warn('⚠️  External mode enabled but .env.external not found');
            }
        }

        // Build unified configuration
        this.config = {
            // Server Configuration
            port: process.env.PORT || 3001,
            nodeEnv: process.env.NODE_ENV || 'development',
            apiBaseUrl: process.env.API_BASE_URL || `http://localhost:${process.env.PORT || 3001}`,
            backendExternalUrl: process.env.BACKEND_EXTERNAL_URL || null,
            
            // External Mode Settings
            externalMode: externalMode,
            frontendUrl: process.env.FRONTEND_EXTERNAL_URL || process.env.CORS_ORIGIN || 'http://localhost:3000',
            
            // Bridge Configuration
            bridgeUrl: process.env.AR_BRIDGE_URL || 'ws://localhost:8765',
            bridgeHttpUrl: process.env.AR_BRIDGE_HTTP_URL || 'http://localhost:8766',
            
            // CORS Configuration
            corsOrigins: this.buildCorsOrigins(),
            
            // WebRTC Configuration
            stunServer: process.env.STUN_SERVER || 'stun:stun.l.google.com:19302',
            turnServer: process.env.TURN_SERVER,
            turnUsername: process.env.TURN_USERNAME,
            turnCredential: process.env.TURN_CREDENTIAL,
            
            // Security
            jwtSecret: process.env.JWT_SECRET,
            rateLimitMax: parseInt(process.env.RATE_LIMIT_MAX_REQUESTS) || (externalMode ? 200 : 100),
            
            // Database
            mongoUri: process.env.MONGODB_URI || 'mongodb://localhost:27017/surgical_platform',
            redisUrl: process.env.REDIS_URL || 'redis://localhost:6379',
            
            // External-specific settings
            externalTimeout: parseInt(process.env.EXTERNAL_TIMEOUT) || 60000,
            connectionRetryAttempts: parseInt(process.env.CONNECTION_RETRY_ATTEMPTS) || 3,
            heartbeatInterval: parseInt(process.env.HEARTBEAT_INTERVAL) || 30000,
            
            // Ngrok settings
            ngrokAuthToken: process.env.NGROK_AUTH_TOKEN,
            ngrokRegion: process.env.NGROK_REGION || 'us',
            ngrokBindTls: process.env.NGROK_BIND_TLS === 'true'
        };
    }

    buildCorsOrigins() {
        const origins = [];

        // Base CORS origins
        const baseOrigins = process.env.CORS_ORIGIN ? 
            process.env.CORS_ORIGIN.split(',').map(o => o.trim()) : 
            ['http://localhost:3000'];
        
        origins.push(...baseOrigins);

        // Add external URL if available
        if (process.env.FRONTEND_EXTERNAL_URL) {
            origins.push(process.env.FRONTEND_EXTERNAL_URL);
        }

        // Add deployment-specific origins
        const deploymentTarget = process.env.REACT_APP_DEPLOYMENT_TARGET || 
                                process.env.DEPLOYMENT_TARGET || 
                                'local';

        switch (deploymentTarget) {
            case 'vercel':
                origins.push('https://*.vercel.app');
                origins.push('https://webrtc-surgical-platform.vercel.app');
                break;
            
            case 'railway':
                origins.push('https://*.railway.app');
                break;
                
            case 'render':
                origins.push('https://*.onrender.com');
                break;
                
            case 'ngrok':
            case 'external':
                origins.push('https://*.ngrok-free.app');
                origins.push('https://*.ngrok.io');
                origins.push('https://*.ngrok.app');
                break;
        }

        // Add development origins
        if (this.config?.nodeEnv === 'development') {
            origins.push('http://localhost:3000');
            origins.push('http://localhost:3001');
            origins.push('http://127.0.0.1:3000');
            origins.push('http://127.0.0.1:3001');
        }

        // Add external mode origins
        if (this.config?.externalMode) {
            origins.push('https://*.ngrok-free.app');
            origins.push('https://*.ngrok.io');
            origins.push('https://*.ngrok.app');
        }

        // Remove duplicates and return unique origins
        const uniqueOrigins = [...new Set(origins)];
        console.log('🔧 CORS Origins configured:', uniqueOrigins);
        
        return uniqueOrigins;
    }

    updateExternalUrls(urls) {
        /**
         * Update configuration with dynamically generated Ngrok URLs
         * @param {Object} urls - Object containing the external URLs
         */
        if (urls.backend) {
            this.config.apiBaseUrl = urls.backend;
            this.config.backendExternalUrl = urls.backend;
            process.env.API_BASE_URL = urls.backend;
            process.env.BACKEND_EXTERNAL_URL = urls.backend;
        }

        if (urls.frontend) {
            this.config.frontendUrl = urls.frontend;
            process.env.FRONTEND_EXTERNAL_URL = urls.frontend;
            process.env.CORS_ORIGIN = urls.frontend;
        }

        if (urls.bridge) {
            this.config.bridgeUrl = urls.bridge;
            process.env.AR_BRIDGE_URL = urls.bridge;
        }

        if (urls.bridgeHttp) {
            this.config.bridgeHttpUrl = urls.bridgeHttp;
            process.env.AR_BRIDGE_HTTP_URL = urls.bridgeHttp;
        }

        // Update CORS origins
        this.config.corsOrigins = this.buildCorsOrigins();

        console.log('🔄 External URLs updated:', {
            backend: this.config.apiBaseUrl,
            frontend: this.config.frontendUrl,
            bridge: this.config.bridgeUrl,
            bridgeHttp: this.config.bridgeHttpUrl
        });
    }

    getConfig() {
        return { ...this.config };
    }

    isExternalMode() {
        return this.config.externalMode;
    }

    getDisplayInfo() {
        const mode = this.isExternalMode() ? 'External' : 'Local';
        return {
            mode,
            backend: this.config.apiBaseUrl,
            frontend: this.config.frontendUrl,
            bridge: this.config.bridgeUrl,
            corsOrigins: this.config.corsOrigins
        };
    }
}

// Singleton instance
const externalConfig = new ExternalConfigManager();

module.exports = externalConfig;